src/dev-server/index.ts

```diff
@@ -0,0 +1,360 @@
+import path from 'path'
+import http from 'http'
+import chalk from 'chalk'
+import stringLength from 'string-length'
+import chokidar from 'chokidar'
+import { formatDistanceToNow } from 'date-fns'
+import { CliUx } from '@oclif/core'
+import { Jamsocket } from '../jamsocket'
+import { buildImage } from '../docker'
+import { SpawnResult, SpawnRequestBody } from '../api'
+
+type Color = 'cyan' | 'magenta' | 'yellow' | 'blue'
+
+type Backend = {
+  spawnResult: SpawnResult,
+  imageId: string,
+  spawnTime: number,
+  lastStatus: string | null,
+  lock: string | null,
+  isStreaming: boolean,
+  color: Color,
+  closeStreams: (() => void) | null,
+}
+
+const BACKEND_LOG_COLORS: Color[] = ['cyan', 'magenta', 'yellow', 'blue']
+
+type Options = {
+  dockerfile: string,
+  service: string,
+  account: string,
+  watch?: string[],
+}
+
+export default class DevServer {
+  currentImageId: string | null = null
+  devBackends: Map<string, Backend> = new Map()
+  curFooterLength = 0
+  totalBackendsSpawned = 0
+
+  constructor(private jamsocket: Jamsocket, private opts: Options) {}
+
+  // returns a promise that resolves when the dev server is stopped
+  async start(): Promise<void> {
+    const { watch } = this.opts
+    this.currentImageId = await this.buildSessionBackend()
+
+    const server = this.startSpawnProxy()
+
+    const timeToExit = new Promise<void>(resolve => {
+      // listen for keyboard input
+      process.stdin.setRawMode(true)
+      process.stdin.resume()
+      process.stdin.setEncoding('utf8')
+      process.stdin.on('data', async (key: string) => {
+        // ctrl-c
```

Comment:
nit: in instances where a comment is used to describe a literal value, I find it slightly nicer to use self-documenting code and use a constant instead, i.e. `const CTRL_C = `\u0003'`

(this is a preference thing, so feel free to ignore if you disagree)

Reply:
Yeah, I think you're right. Will update!